internal/util/ssa/managedfields.go

```diff
@@ -20,86 +20,197 @@ package ssa
-	// This will ensure that `ssaManager` will be able to modify the fields that
-	// were originally owned by "manager".
-	base := obj.DeepCopyObject().(client.Object)
-
-	// Modify managedFieldEntry for manager=manager and operation=update to drop ownership
-	// for the given paths to avoid having two managers holding values.
-	originalManagedFields := obj.GetManagedFields()
+	base := object.DeepCopyObject().(client.Object)
+
+	// Remove managedFields for fieldManager:Apply and manager:Update.
+	originalManagedFields := object.GetManagedFields()
 	managedFields := make([]metav1.ManagedFieldsEntry, 0, len(originalManagedFields))
 	for _, managedField := range originalManagedFields {
+		if managedField.Manager == fieldManager &&
+			managedField.Operation == metav1.ManagedFieldsOperationApply &&
+			managedField.Subresource == "" {
+			continue
+		}
 		if managedField.Manager == classicManager &&
-			managedField.Operation == metav1.ManagedFieldsOperationUpdate {
+			managedField.Operation == metav1.ManagedFieldsOperationUpdate &&
+			managedField.Subresource == "" {
+			continue
+		}
+		managedFields = append(managedFields, managedField)
+	}
+
+	// Add a seeding managedField entry to prevent SSA to create/infer a default managedField entry when the
+	// first SSA is applied.
+	// If an existing object doesn't have managedFields when applying the first SSA the API server
+	// creates an entry with operation=Update (guessing where the object comes from), but this entry ends up
+	// acting as a co-ownership and we want to prevent this.
+	// NOTE: fieldV1Map cannot be empty, so we add metadata.name which will be cleaned up at the first
+	//       SSA patch of the same manager (updateLabelsAndAnnotations directly after calling this func).
+	managedFields = append(managedFields, metav1.ManagedFieldsEntry{
+		Manager:    metadataFieldManager,
+		Operation:  metav1.ManagedFieldsOperationApply,
+		APIVersion: objectGVK.GroupVersion().String(),
+		Time:       ptr.To(metav1.Now()),
+		FieldsType: "FieldsV1",
+		FieldsV1:   &metav1.FieldsV1{Raw: []byte(`{"f:metadata":{"f:name":{}}}`)},
+	})
+
+	object.SetManagedFields(managedFields)
+
+	// Use optimistic locking to avoid accidentally rolling back managedFields.
+	if err := c.Patch(ctx, object, client.MergeFromWithOptions(base, client.MergeFromWithOptimisticLock{})); err != nil {
+		return errors.Wrapf(err, "failed to migrate managedFields for %s %s",
+			objectGVK.Kind, klog.KRef(objectKey.Namespace, objectKey.Name))
+	}
+	return nil
+}
+
+// needsMigration returns true if fieldManager:Apply owns the clusterv1.ClusterNameLabel.
+func needsMigration(object client.Object, fieldManager string) (bool, error) {
+	for _, managedField := range object.GetManagedFields() {
+		//nolint:gocritic // much easier to read this way, not going to use continue instead
+		if managedField.Manager == fieldManager &&
+			managedField.Operation == metav1.ManagedFieldsOperationApply &&
+			managedField.Subresource == "" {
```

Comment:
Wondering if we can avoid Marshal/UnMarshal by adding the following condition

```suggestion
		if managedField.Manager == fieldManager &&
			managedField.Operation == metav1.ManagedFieldsOperationApply &&
			managedField.Subresource == ""  && bytes.Contains(managedField.FieldsV1.Raw, []byte("f:cluster.x-k8s.io/cluster-name")) {
```

(or using a regEx for a more sophisticated check that ensure "f:cluster.x-k8s.io/cluster-name" is below metadata)